File: data/src/main/java/com/github/anrimian/musicplayer/data/storage/files/StorageFilesDataSource.java
package com.github.anrimian.musicplayer.data.storage.files;

import androidx.core.util.Pair;

import com.github.anrimian.musicplayer.data.storage.providers.music.FilePathComposition;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.composition.FullComposition;

import java.util.List;

public interface StorageFilesDataSource {
    String renameCompositionsFolder(List<Composition> compositions,
                                    String oldPath,
                                    String newName,
                                    List<FilePathComposition> updatedCompositions);

    Pair<String, String> renameCompositionFile(FullComposition composition, String fileName);

    List<FilePathComposition> moveCompositionsToFolder(List<Composition> compositions,
                                                       String fromPath,
                                                       String toPath);

    String moveCompositionsToNewFolder(List<Composition> compositions,
                                       String fromPath,
                                       String toParentPath,
                                       String directoryName,
                                       List<FilePathComposition> updatedCompositions);

    List<Composition> deleteCompositionFiles(List<Composition> compositions, Object tokenForDelete);

    void deleteCompositionFile(Composition composition);

    void clearDeleteData();
}

File: data/src/main/java/com/github/anrimian/musicplayer/data/storage/files/StorageFilesDataSource.kt
package com.github.anrimian.musicplayer.data.storage.files

import com.github.anrimian.musicplayer.domain.models.composition.DeletedComposition

interface StorageFilesDataSource {

    /**
     * if dest folder already exist - files will be merged
     * returns id-new_file_name - if name was changed after rename
     */
    fun renameCompositionsDirectory(
        files: List<StorageFileInfo>,
        folderPath: String,
        newFolderPath: String
    ): List<Pair<Long, String>>

    fun renameCompositionFile(storageId: Long, fileName: String): String

    fun moveCompositionsToDirectory(
        files: Collection<StorageFileInfo>,
        fromFolderPath: String,
        toFolderPath: String
    ): List<Pair<Long, String>>

    fun executePathChange(moveOperations: List<StorageMoveOperation>): List<Pair<Long, String>>

    fun deleteCompositionFiles(
        compositions: List<DeletedComposition>,
        tokenForDelete: Any
    ): List<DeletedComposition>

    fun deleteCompositionFile(composition: DeletedComposition): DeletedComposition

    fun clearDeleteData()

}